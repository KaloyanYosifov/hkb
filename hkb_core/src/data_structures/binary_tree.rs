use std::{cell::RefCell, rc::Rc};

trait Constraints: PartialEq + Eq + PartialOrd + Ord {}

impl<T: PartialEq + Eq + PartialOrd + Ord> Constraints for T {}

pub type NodeRef<T> = Rc<RefCell<Node<T>>>;

#[derive(Ord, Eq, Debug)]
pub struct Node<T: PartialEq + Eq + PartialOrd + Ord> {
    pub val: T,
    left: Option<NodeRef<T>>,
    right: Option<NodeRef<T>>,
}

impl<T: Constraints> PartialEq for Node<T> {
    fn eq(&self, other: &Self) -> bool {
        self.val == other.val
    }
}

impl<T: Constraints> PartialOrd for Node<T> {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        let val = {
            if self.val > other.val {
                std::cmp::Ordering::Greater
            } else if self.val < other.val {
                std::cmp::Ordering::Less
            } else {
                std::cmp::Ordering::Equal
            }
        };

        Some(val)
    }
}

impl<T: Constraints> Node<T> {
    pub fn new(val: T, left_val: Option<T>, right_val: Option<T>) -> Self {
        let left = {
            if let Some(v) = left_val {
                Some(Rc::new(RefCell::new(Node::with_value(v))))
            } else {
                None
            }
        };
        let right = {
            if let Some(v) = right_val {
                Some(Rc::new(RefCell::new(Node::with_value(v))))
            } else {
                None
            }
        };

        Self { val, left, right }
    }

    pub fn with_value(val: T) -> Self {
        Self {
            val,
            left: None,
            right: None,
        }
    }

    pub fn with_nodes(val: T, left: Node<T>, right: Node<T>) -> Self {
        Self {
            val,
            left: Some(Rc::new(RefCell::new(left))),
            right: Some(Rc::new(RefCell::new(right))),
        }
    }

    pub fn get_left<'a>(&self) -> Option<NodeRef<T>> {
        if let Some(node) = self.left.as_ref() {
            Some(node.clone())
        } else {
            None
        }
    }

    pub fn get_right<'a>(&self) -> Option<NodeRef<T>> {
        if let Some(node) = self.right.as_ref() {
            Some(node.clone())
        } else {
            None
        }
    }

    pub fn height(&self) -> usize {
        self.height_recursive(self)
    }

    fn height_recursive(&self, node: &Node<T>) -> usize {
        match (node.get_left(), node.get_right()) {
            (Some(left), Some(right)) => {
                let left_height = self.height_recursive(&left.borrow());
                let right_height = self.height_recursive(&right.borrow());
                1 + std::cmp::max(left_height, right_height)
            }
            (Some(left), None) => 1 + self.height_recursive(&left.borrow()),
            (None, Some(right)) => 1 + self.height_recursive(&right.borrow()),
            (None, None) => 1,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Node;

    #[test]
    fn it_can_create_a_node_with_value() {
        let node = Node::with_value(3);

        assert_eq!(3, node.val);
        assert!(matches!(node.get_left(), None));
        assert!(matches!(node.get_right(), None));
    }

    #[test]
    fn it_can_create_a_node_with_left_right_moved() {
        let node = Node::new(3, Some(5), Some(7));

        assert_eq!(3, node.val);
        assert_eq!(5, node.get_left().unwrap().borrow().val);
        assert_eq!(7, node.get_right().unwrap().borrow().val);
    }

    #[test]
    fn it_can_create_with_nodes_moved_directly() {
        let left_node = Node::with_value(5);
        let right_node = Node::with_value(7);
        let node = Node::with_nodes(3, left_node, right_node);

        assert_eq!(3, node.val);
        assert_eq!(5, node.get_left().unwrap().borrow().val);
        assert_eq!(7, node.get_right().unwrap().borrow().val);
    }

    #[test]
    fn it_can_return_the_height_of_the_node_when_it_is_one() {
        let node = Node::with_value(3);

        assert_eq!(1, node.height());
    }

    #[test]
    fn it_can_return_the_height_of_the_node() {
        let left_node = Node::with_value(5);
        let right_node = Node::with_value(7);
        let node = Node::with_nodes(3, left_node, right_node);

        assert_eq!(2, node.height());
    }
}
