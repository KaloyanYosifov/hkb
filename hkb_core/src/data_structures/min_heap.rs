/// Min heap
/// 1. Always the most minimum value is on top
/// 2. We have a trait which returns an integer value
/// 3. The trait will be dynamic and boxed
///
/// Implementation
/// 1. Insert:
///     - Put the value in the end
///     - Bubble up
/// 2. Pop
///     - Keep the value we are going to return
///     - Swap with the bottom most element
///     - remove it from the last element
///     - Start from root and heapify to fix the top minimum value

// TODO:
// Maybe we can do traits support here and a user can handle any internal obj they want to keep
// Something like MinHeapValue trait and the user can implement a struct with this trait and the
// struct can have { node: Some(node), node_val: i64, somehting: None }

pub struct MinHeap {
    values: Vec<i64>,
}

impl MinHeap {
    pub fn new() -> Self {
        Self::with_capacity(32)
    }

    pub fn with_capacity(capacity: usize) -> Self {
        Self {
            values: Vec::with_capacity(capacity),
        }
    }
}

impl MinHeap {
    pub fn insert(&mut self, value: i64) {
        self.values.push(value);

        self.heapify_bottom_up();
    }

    pub fn pop(&mut self) -> Option<i64> {
        if self.values.len() == 0 {
            None
        } else if self.values.len() == 1 {
            self.values.pop()
        } else {
            self.swap(0, self.values.len() - 1);

            let value = self.values.pop();

            self.heapify_top_down();

            value
        }
    }

    fn heapify_bottom_up(&mut self) {
        let mut index = self.values.len() - 1;
        let mut parent = index / 2;

        while self.values[parent] > self.values[index] {
            self.swap(index, parent);
            index = parent;
            parent = index / 2;
        }
    }

    fn heapify_top_down(&mut self) {
        let mut index = 0;

        loop {
            let current_val = self.values[index];
            let left = (index * 2) + 1;
            let right = (index * 2) + 2;

            if left < self.values.len() && right < self.values.len() {
                let left_val = self.values[left];
                let right_val = self.values[right];

                if left_val < right_val && left_val < current_val {
                    self.swap(left, index);
                    index = left;
                } else if right_val < left_val && right_val < current_val {
                    self.swap(right, index);
                    index = right;
                } else {
                    break;
                }
            } else if left < self.values.len() {
                let left_val = self.values[left];

                if left_val < current_val {
                    self.swap(left, index);
                    index = left;
                } else {
                    break;
                }
            } else if right < self.values.len() {
                let right_val = self.values[right];

                if right_val < current_val {
                    self.swap(right, index);
                    index = right;
                } else {
                    break;
                }
            } else {
                break;
            }
        }
    }

    fn swap(&mut self, index1: usize, index2: usize) {
        let temp = self.values[index1];
        self.values[index1] = self.values[index2];
        self.values[index2] = temp;
    }
}

#[cfg(test)]
mod tests {
    use super::MinHeap;

    // TODO: add property testing

    fn create_min_heap() -> MinHeap {
        let mut min_heap = MinHeap::new();
        let mut values: [usize; 10] = core::array::from_fn(|i| i + 1);

        values.reverse();

        for i in values {
            min_heap.insert(i as i64);
        }

        min_heap
    }

    #[test]
    fn it_can_insert_a_value() {
        let mut min_heap = MinHeap::new();

        min_heap.insert(5);

        assert_eq!(5, min_heap.values[0]);
    }

    #[test]
    fn it_puts_the_min_value_at_top() {
        let mut min_heap = MinHeap::new();
        let mut values: [usize; 10] = core::array::from_fn(|i| i + 1);

        values.reverse();

        for i in values {
            min_heap.insert(i as i64);
            assert_eq!(i as i64, min_heap.values[0]);
        }

        assert_eq!(1, min_heap.values[0]);
    }

    #[test]
    fn it_does_not_put_new_high_value_to_the_top() {
        let mut min_heap = create_min_heap();

        min_heap.insert(300);

        assert_eq!(1, min_heap.values[0]);
    }

    #[test]
    fn it_can_get_top_value_from_min_heap() {
        let mut min_heap = create_min_heap();

        let val = min_heap.pop();

        assert_eq!(1, val.unwrap());
        assert_eq!(2, min_heap.values[0]);
    }
}
