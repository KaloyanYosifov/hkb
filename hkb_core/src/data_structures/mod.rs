pub mod binary_tree;
pub mod min_heap;
