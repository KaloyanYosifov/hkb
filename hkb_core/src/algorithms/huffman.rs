use std::collections::HashMap;

use crate::data_structures::Node;

#[derive(Eq, Ord, Debug)]
pub struct HuffmanValue {
    char: char,
    occurance: u64,
}

impl PartialEq for HuffmanValue {
    fn eq(&self, other: &Self) -> bool {
        self.char == other.char
    }
}

impl PartialOrd for HuffmanValue {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        let val = {
            if self.occurance > other.occurance {
                std::cmp::Ordering::Greater
            } else if self.occurance < other.occurance {
                std::cmp::Ordering::Less
            } else {
                std::cmp::Ordering::Equal
            }
        };

        Some(val)
    }
}

pub struct HuffmanEncoder;

impl HuffmanEncoder {
    pub fn compress(text: &str) -> Node<HuffmanValue> {
        let mut huffman_values: Vec<HuffmanValue> = Vec::with_capacity(text.len() / 2);
        let mut occurance_map: HashMap<char, usize> = HashMap::with_capacity(text.len() / 2);

        for c in text.chars().into_iter() {
            let index = occurance_map
                .entry(c)
                .or_insert_with(|| huffman_values.len() + 1);
            let actual_index = *index - 1;

            if actual_index >= huffman_values.len() {
                huffman_values.push(HuffmanValue {
                    char: c,
                    occurance: 0,
                });
            }

            huffman_values[actual_index].occurance += 1;
        }

        huffman_values.sort();

        println!("{:?}", huffman_values);

        let node = Node::with_value(HuffmanValue {
            char: 'a',
            occurance: 32,
        });

        node
    }
}

#[cfg(test)]
mod tests {
    use super::HuffmanEncoder;

    #[test]
    fn it_can_create_huffman_table_from_text() {
        let text = "Hello there magnificent mothertrucker";

        let node = HuffmanEncoder::compress(text);

        assert!(false);
    }
}
