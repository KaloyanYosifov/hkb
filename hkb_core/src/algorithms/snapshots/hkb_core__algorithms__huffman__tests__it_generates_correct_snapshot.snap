---
source: hkb_core/src/algorithms/huffman.rs
expression: "HuffmanEncoder::compress(text)"
---
(
    HuffmanBinary {
        binary: [
            HuffmanBinaryValue {
                val: 10,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 31,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 31,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 14,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 4,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 0,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 1,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 14,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 4,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 11,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 2,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 30,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 10,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 3,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 10,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 11,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 30,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 4,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 14,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 4,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 4,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 0,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 1,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 4,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 1,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 14,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 11,
                max_bits: 4,
            },
            HuffmanBinaryValue {
                val: 15,
                max_bits: 5,
            },
            HuffmanBinaryValue {
                val: 6,
                max_bits: 3,
            },
            HuffmanBinaryValue {
                val: 1,
                max_bits: 3,
            },
        ],
    },
    Node {
        val: HuffmanValue {
            char: None,
            occurance: 37,
        },
        left: Some(
            RefCell {
                value: Node {
                    val: HuffmanValue {
                        char: None,
                        occurance: 16,
                    },
                    left: Some(
                        RefCell {
                            value: Node {
                                val: HuffmanValue {
                                    char: None,
                                    occurance: 8,
                                },
                                left: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: None,
                                                occurance: 4,
                                            },
                                            left: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: Some(
                                                                'h',
                                                            ),
                                                            occurance: 2,
                                                        },
                                                        left: None,
                                                        right: None,
                                                    },
                                                },
                                            ),
                                            right: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: None,
                                                            occurance: 2,
                                                        },
                                                        left: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'g',
                                                                        ),
                                                                        occurance: 1,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                        right: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'f',
                                                                        ),
                                                                        occurance: 1,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                    },
                                                },
                                            ),
                                        },
                                    },
                                ),
                                right: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: Some(
                                                    'r',
                                                ),
                                                occurance: 4,
                                            },
                                            left: None,
                                            right: None,
                                        },
                                    },
                                ),
                            },
                        },
                    ),
                    right: Some(
                        RefCell {
                            value: Node {
                                val: HuffmanValue {
                                    char: None,
                                    occurance: 8,
                                },
                                left: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: None,
                                                occurance: 4,
                                            },
                                            left: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: Some(
                                                                'm',
                                                            ),
                                                            occurance: 2,
                                                        },
                                                        left: None,
                                                        right: None,
                                                    },
                                                },
                                            ),
                                            right: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: None,
                                                            occurance: 2,
                                                        },
                                                        left: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'H',
                                                                        ),
                                                                        occurance: 1,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                        right: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'a',
                                                                        ),
                                                                        occurance: 1,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                    },
                                                },
                                            ),
                                        },
                                    },
                                ),
                                right: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: None,
                                                occurance: 4,
                                            },
                                            left: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: Some(
                                                                'o',
                                                            ),
                                                            occurance: 2,
                                                        },
                                                        left: None,
                                                        right: None,
                                                    },
                                                },
                                            ),
                                            right: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: None,
                                                            occurance: 2,
                                                        },
                                                        left: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'u',
                                                                        ),
                                                                        occurance: 1,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                        right: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'k',
                                                                        ),
                                                                        occurance: 1,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                    },
                                                },
                                            ),
                                        },
                                    },
                                ),
                            },
                        },
                    ),
                },
            },
        ),
        right: Some(
            RefCell {
                value: Node {
                    val: HuffmanValue {
                        char: None,
                        occurance: 21,
                    },
                    left: Some(
                        RefCell {
                            value: Node {
                                val: HuffmanValue {
                                    char: None,
                                    occurance: 8,
                                },
                                left: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: Some(
                                                    't',
                                                ),
                                                occurance: 4,
                                            },
                                            left: None,
                                            right: None,
                                        },
                                    },
                                ),
                                right: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: None,
                                                occurance: 4,
                                            },
                                            left: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: Some(
                                                                'i',
                                                            ),
                                                            occurance: 2,
                                                        },
                                                        left: None,
                                                        right: None,
                                                    },
                                                },
                                            ),
                                            right: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: Some(
                                                                'c',
                                                            ),
                                                            occurance: 2,
                                                        },
                                                        left: None,
                                                        right: None,
                                                    },
                                                },
                                            ),
                                        },
                                    },
                                ),
                            },
                        },
                    ),
                    right: Some(
                        RefCell {
                            value: Node {
                                val: HuffmanValue {
                                    char: None,
                                    occurance: 13,
                                },
                                left: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: Some(
                                                    'e',
                                                ),
                                                occurance: 6,
                                            },
                                            left: None,
                                            right: None,
                                        },
                                    },
                                ),
                                right: Some(
                                    RefCell {
                                        value: Node {
                                            val: HuffmanValue {
                                                char: None,
                                                occurance: 7,
                                            },
                                            left: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: Some(
                                                                ' ',
                                                            ),
                                                            occurance: 3,
                                                        },
                                                        left: None,
                                                        right: None,
                                                    },
                                                },
                                            ),
                                            right: Some(
                                                RefCell {
                                                    value: Node {
                                                        val: HuffmanValue {
                                                            char: None,
                                                            occurance: 4,
                                                        },
                                                        left: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'n',
                                                                        ),
                                                                        occurance: 2,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                        right: Some(
                                                            RefCell {
                                                                value: Node {
                                                                    val: HuffmanValue {
                                                                        char: Some(
                                                                            'l',
                                                                        ),
                                                                        occurance: 2,
                                                                    },
                                                                    left: None,
                                                                    right: None,
                                                                },
                                                            },
                                                        ),
                                                    },
                                                },
                                            ),
                                        },
                                    },
                                ),
                            },
                        },
                    ),
                },
            },
        ),
    },
)
